### PR TITLE
Bls provider json rpc provider methods

### LINKED ISSUE #381: BlsProvider JsonRpcProvider Methods
We are adding a BlsProvider to `bls-wallet-clients`. This issue is concerned with ensuring the JsonRpcProvider Methods work with the BlsProvider. This will include overriding methods with BLS logic if needed and testing the other methods work as expected.

Some of the operations include getting the signer, listing accounts and sending raw messages.

[JsonRpcProvider Methods](https://docs.ethers.io/v5/api/providers/jsonrpc-provider/) to test/override:

```ts
// Connect to a JSON-RPC HTTP API using the URL or ConnectionInfo urlOrConnectionInfo 
// connected to the networkish network.
new ethers.providers.JsonRpcProvider( [ urlOrConnectionInfo [ , networkish ] ] )

// The fully formed ConnectionInfo the Provider is connected to.
jsonRpcProvider.connection ⇒ ConnectionInfo

// Returns a JsonRpcSigner which is managed by this Ethereum node, at addressOrIndex. If no 
// addressOrIndex is provided, the first account (account #0) is used.
jsonRpcProvider.getSigner( [ addressOrIndex ] ) ⇒ JsonRpcSigner

// No description given in docs
jsonRpcProvider.getUncheckedSigner( [ addressOrIndex ] ) ⇒ JsonRpcUncheckedSigner

// Returns a list of all account addresses managed by this provider.
jsonRpcProvider.listAccounts( ) ⇒ Promise< Array< string > >

// Allows sending raw messages to the provider.
jsonRpcProvider.send( method , params ) ⇒ Promise< any >
```

### PR DESCRIPTION
**Pull request by JohnGuilding**
## What is this PR doing?
This PR is concerned with adding/testing the [JsonRpcProvider methods](https://docs.ethers.io/v5/api/providers/jsonrpc-provider/#JsonRpcProvider).

Relevant methods:

```ts
jsonRpcProvider.connection
getSigner()
getUncheckedSigner()
listAccounts()
send()
```

**Notes:**

**getSigner()** - was needed in previous work, so implementation and tests have already been added.

**getUncheckedSigner()** - in order to test this method fully, I had to add several additional methods to the signer along with tests to ensure the unchecked signer worked as expected. This includes adding the `UncheckedBlsSigner` class itself. Methods added/changed:

- `signer.connectBlsUnchecked()` - see [comment](https://github.com/web3well/bls-wallet/pull/422#discussion_r1044535148) on why I added this
- `signer.sendUncheckedTransaction()`
- `uncheckedBlsSigner.sendTransaction()`

For more information on what the unchecked signer is and why it was added to ethers see this [part of the docs](https://docs.ethers.io/v5/api/providers/jsonrpc-provider/#UncheckedJsonRpcSigner) and this [github issue](https://github.com/ethers-io/ethers.js/issues/340)

**send()** - added some tests here to ensure we called the underlying ethers logic, and got the expected results for some common rpc requests. See [this comment](https://github.com/web3well/bls-wallet/pull/422/files#r1044535148) for further discussion.

## How can these changes be manually tested?
1. run local hardhat node
2. Spin up a local aggregator
3. in the `./contracts` directory, run `yarn test-clients`

## Does this PR resolve or contribute to any issues?
#381  

## Checklist
- [x] I have manually tested these changes
- [x] Post a link to the PR in the group chat

## Guidelines
- If your PR is not ready, mark it as a draft
- The `resolve conversation` button is for reviewers, not authors
  - (But add a 'done' comment or similar)

### CODE REVIEW

File: contracts/clients-tests/BlsProvider.test.ts [hunk 1]
```diff
@@ -428,6 +439,60 @@ describe("BlsProvider", () => {
...
+    const expectedAccounts = await regularProvider.listAccounts();
+
+    // Act
+    const accounts = await blsProvider.listAccounts();
+
+    // Assert
+    expect(accounts).to.deep.equal(expectedAccounts);
+  });
+
+  it("should send an rpc request to the provider", async () => {
```
Comment (JohnGuilding) [inline contracts/clients-tests/BlsProvider.test.ts:468]: Undecided on further changes here and whether we should override/extend (not sure what correct terminology is here) the rpc calls to execute bls logic instead, maybe similar to [how we've done in the QuillProvider](https://github.com/web3well/bls-wallet/blob/4f5b7c75d9cf2baef9e44469f016a89d819df4d2/extension/source/background/AggregatorController.ts#L38-L97)?. The overarching question to answer here is **what should the expected behaviour be when a developer uses `blsProvider.send()`?**

File: contracts/clients/src/BlsSigner.ts [hunk 2]
```diff
@@ -200,17 +200,56 @@ export default class BlsSigner extends Signer {
...
-  connectUnchecked(): ethers.providers.JsonRpcSigner {
-    throw new Error("connectUnchecked() is not implemented.");
+  connectUnchecked(): BlsSigner {
+    throw new Error(
+      "connectUnchecked() is not implemented. Use connectBlsUnchecked instead.",
+    );
+  }
+
+  // When we use regular "connectUnchecked", we need to pass the private key into the method as that is a required
+  // argument for the BlsSigner constructor. This breaks liskov substitution as the subclass method signature is different.
```
Comment (JohnGuilding) [inline contracts/clients/src/BlsSigner.ts]: In a previous PR, there was an agreement that adding custom methods to be used in place of an ethers method was undesirable. So I've created this method apprehensively, but I'm unsure how we could get `connectUnchecked` to work due to needing a different method signature. Open to suggestions on better ways to resolve this.

File: contracts/clients-tests/BlsProvider.test.ts (hunk 1 above)
Comment (jacque006) [reply contracts/clients-tests/BlsProvider.test.ts:468]: I would expect this to work similarly to the `QuillProvider`/be BLS Wallet specific when:
- Sending a transaction
- Sending multiple transactions (currently overriding/extending `eth_sendTransaction`, we really should switch to a custom RPC namespace like `blsWallet_sendTransactions`).
- Looking up a receipt from a BLS Wallet txn/bundle

And like a normal provider for other send calls.

We actually can't send/sign normal EOA transactions & messages (ECDSA) as we only have our BLS pub/priv keys. We *may* want to add that in the future if:
- Someone wants to bypass an aggregator and submit straight to chain
- Someone wants to use this provider to write a custom aggregator
- We want to use this in our own aggregator

But I think those are out of scope of this initial implementation and we should focus mostly on use cases where a dApp/script/service want an embedded BLS burner wallet, similar to the wallet @blakecduncan created.

File: contracts/clients/src/BlsSigner.ts [hunk 3]
```diff
@@ -200,17 +200,56 @@ export default class BlsSigner extends Signer {
...
+  connectBlsUnchecked(privateKey: string): BlsSigner {
+    return new UncheckedBlsSigner(
+      _constructorGuard,
+      this.provider,
+      privateKey,
+      this._address || this._index,
+    );
```
Comment (jacque006) [inline contracts/clients/src/BlsSigner.ts]:
```suggestion
  connectUnchecked(): BlsSigner {
    return new UncheckedBlsSigner(
      _constructorGuard,
      this.provider,
      // Modify this arg to take a priv key OR promise that will eventually resolve to one, then do similar initPromise guards in this class.
      this.wallet?.privateKey ?? async () => {
      	await this.initPromise();
      	return this.wallet.privateKey;
      },
      this._address || this._index,
    );
  }
```

File: contracts/clients/src/BlsSigner.ts (hunk 2 above)
Comment (jacque006) [reply contracts/clients/src/BlsSigner.ts]: https://github.com/web3well/bls-wallet/pull/422/files#r1044853129

File: contracts/clients/src/BlsSigner.ts (hunk 3 above)
Comment (jacque006) [reply contracts/clients/src/BlsSigner.ts]: This requires modifying the `BlsSigner` constructor to accept a `privateKey` or a promise that will eventually resolve to one. We can discuss offline some cleaner ways to do this.

File: contracts/clients-tests/BlsProvider.test.ts (hunk 1 above)
Comment (JohnGuilding) [reply contracts/clients-tests/BlsProvider.test.ts:468]: Thanks for the thoughts. Makes sense regarding expected behaviour. 

Thinking the same that this work is out of scope for initial implementation. Have raised an issue to look at this in the future (out of scope for current milestone): #424

File: contracts/clients-tests/BlsSigner.test.ts [hunk 4]
```diff
@@ -370,6 +371,137 @@ describe("BlsSigner", () => {
...
+
+    // Assert
+    expect(
+      (await blsProvider.getBalance(recipient)).sub(balanceBefore),
+    ).to.equal(transactionAmount);
+  });
+
+  it("should throw an error getting the transaction receipt when using a new provider and connecting to an unchecked bls signer", async () => {
+    // Arrange & Act
+    // This test is identical to the above test except this one uses a new instance of a provider, yet fails to find the tx receipt
```
Comment (JohnGuilding) [inline contracts/clients-tests/BlsSigner.test.ts]: As described in the comment, this test fails to find the tx receipt when instantiating the unchecked signer from a new provider instance. Not sure why this is happening?

File: contracts/clients/src/BlsSigner.ts [hunk 5]
```diff
@@ -200,17 +201,55 @@ export default class BlsSigner extends Signer {
     throw new Error("_signTypedData() is not implemented.");
   }
 
-  connectUnchecked(): ethers.providers.JsonRpcSigner {
-    throw new Error("connectUnchecked() is not implemented.");
+  connectUnchecked(): BlsSigner {
+    return new UncheckedBlsSigner(
+      _constructorGuard,
+      this.provider,
+      // Modify this arg to take a priv key OR promise that will eventually resolve to one, then do similar initPromise guards in this class.
```
Comment (jacque006) [inline contracts/clients/src/BlsSigner.ts]: I included this originally to illustrate my suggestion, not needed in code IMO

```suggestion
```

File: contracts/clients-tests/BlsSigner.test.ts (hunk 4 above)
Comment (jacque006) [reply contracts/clients-tests/BlsSigner.test.ts]: https://github.com/web3well/bls-wallet/issues/427

**Review by jacque006 — APPROVED** [review]

**Pull request merged**